build: add basic devcontainer.json (for Rust, JavaScript, Python)

eitsupi opened the pull request:
Adresses #1808

eitsupi committed:
chore: add devcontainer config file

max-sixty committed:
Merge branch 'main' into devcontainer

eitsupi committed:
Merge remote-tracking branch 'upstream/main' into devcontainer

eitsupi commented on .github/workflows/devcontainer.yaml:
```diff
@@ -0,0 +1,47 @@
+      - name: Pre-build dev container image
+        uses: devcontainers/ci@v0.2
+        with:
+          subFolder: .devcontainer/base
```
It does not seem to be possible to configure this way; I have created an issue.
devcontainers/ci#212

max-sixty replied:
Hmmm. I wonder whether this is substantially better than using a normal docker workflow like we have now. Because I _think_ the only thing that's pushed is the output of the Docker build. Is that right?

If it is, it might be that we can mostly swap out the existing Dockerfile for the new one, and use the existing workflow...

eitsupi replied:
> I wonder whether this is substantially better than using a normal docker workflow like we have now. Because I _think_ the only thing that's pushed is the output of the Docker build. Is that right?

Since Dev Container Features are a fairly long shell script, I think it provides a better setup than simply installing Python, etc.
Dev Container CLI is required to build a Docker image with Dev Container Features installed.

But as you point out, it doesn't support all the features of `buildx build` (such as GitHub Actions cache support), so it's worth moving to a simple Docker file-only build.
Is that better for you?
(In that case, change the installation of Dev Container Features to be done at the user's end. This takes some time, but is far more minimal than `cargo install`.)

max-sixty replied:
Oh I think I see now — this will cache running the steps in `.devcontainer/devcontainer.json` as well as the Dockerfile? That's excellent if so!

~In that case I would probably keep it as-is, because then folks don't need to do any building themselves, they get a fully baked image...~ Writing down my understanding below

eitsupi committed:
build: add base devcontainer definition and build workflow file

eitsupi commented on .github/workflows/devcontainer.yaml:
```diff
@@ -0,0 +1,53 @@
+        with:
+          subFolder: .devcontainer/base
+          imageName: ghcr.io/prql/prql-devcontainer
+          cacheFrom: ghcr.io/prql/prql-devcontainer
```
Is this a good name for the image to push?
Or like `ghcr.io/prql/prql-base-devcontainer`?

max-sixty replied:
Yes I think it's great

eitsupi commented on .github/workflows/devcontainer.yaml:
```diff
@@ -0,0 +1,53 @@
+        with:
+          subFolder: .devcontainer/base
+          imageName: ghcr.io/prql/prql-devcontainer
+          cacheFrom: ghcr.io/prql/prql-devcontainer
```
> it's much easier than this for the basic Docker caching — this does the basic build cache already, using the image repository:

Inline cache is used here.

max-sixty replied:
Yes great! That's all the caching we need; no need for the build mount

richb-hanover committed:
First cut, very low-quality draft for Dev Containers

I have written about 98% of what I know. Let's all contribute to strengthen this document. Thanks.

richb-hanover committed:
Remove "will" in favor of direct action words.

It's almost always better to write documentation saying "thing X _does_ action Y" instead of "thing X will do action Y".

richb-hanover committed:
Update using-dev-container.md

richb-hanover committed:
Update using-dev-container.md

richb-hanover committed:
Final editorial tweak for first-cut description

max-sixty committed:
Merge branch 'main' into devcontainer

eitsupi committed:
Merge branch 'main' into devcontainer

richb-hanover committed:
Update to reflect @eitsupi's comments

eitsupi committed:
docs: First cut draft for Dev Containers

pre-commit-ci[bot] committed:
[pre-commit.ci] auto fixes from pre-commit.com hooks

for more information, see https://pre-commit.ci

max-sixty commented on book/src/contributing/using-dev-container.md:
```diff
@@ -0,0 +1,68 @@
+
+## Starting and stopping the Dev Container
+
+_Instructions needed - both from VS Code and maybe from CLI?._
```
I would generally deemphasize documenting anything that isn't PRQL-specific, such as how to use a dev container, and instead just link to the official documentation. It's often not as good as the official version, difficult to maintain, and accrues debt that's difficult to assess.

That said, I think this is well-written, so we don't have to remove it, but I wouldn't invest more.

eitsupi replied:
For the record, this document is entirely the credit of @richb-hanover.
Sorry, but I have very less hand in the documentation yet.

I agree that it is not necessary to explain here how to use Dev Container in general.
And, this document does not sufficiently distinguish Dev Container from VS Code Dev Container (extension) and I need to minor correct this.......

For example, https://github.com/pandas-dev/pandas has included devcontainer.json for several years, and described as this.
https://pandas.pydata.org/docs/development/contributing_environment.html?highlight=devcontainer#id5

max-sixty replied:
> For the record, this document is entirely the credit of @richb-hanover.

Ah congrats @richb-hanover ! It is very well-written!

> For example, https://github.com/pandas-dev/pandas has included devcontainer.json for several years, and described as this.
https://pandas.pydata.org/docs/development/contributing_environment.html?highlight=devcontainer#id5

Yes I think that's about the right amount of explanation...

max-sixty commented on book/src/contributing/using-dev-container.md:
```diff
@@ -0,0 +1,68 @@
+
+When you first start the Dev Container, the build process may take a long time
+(as much as 20-40 minutes) as the container collects all the packages of the
+full PRQL toolchain. Once it's running, though, startup is fairly fast.
```
This part is helpful PRQL-specific context though

max-sixty commented on book/src/contributing/using-dev-container.md:
```diff
@@ -0,0 +1,68 @@
+  edit the files (in the `playground` directory), the server rebuilds those
+  pages. (Port 3000)
+
+- `task WHAT ELSE?` _Provide explanation of other useful commands._
```
I would just link to our existing docs here

max-sixty commented on book/src/contributing/using-dev-container.md:
```diff
@@ -0,0 +1,68 @@
+install the (single) Dev Container on their own machine to start working. By
+definition, the Dev Container has a consistent set of tools that are known to
+work together. This avoids a fuss with finding the proper versions of each of
+the build tools.
```
This is still a reasonable summary to include, it's very well-written!

max-sixty commented on book/src/contributing/using-dev-container.md:
```diff
@@ -0,0 +1,68 @@
+   Container watch for changed files and rebuild the project as needed.
+
+3. When you are satisfied with the changes, you can commit them to the repo, and
+   push the changes as usual.
```
I probably wouldn't keep this, though. But fine to leave it in for the moment and see how we go

eitsupi committed:
Merge branch 'main' into devcontainer

eitsupi commented on book/src/contributing/using-dev-container.md:
```diff
@@ -0,0 +1,68 @@
+  extension for VS Code
+
+When you first start the Dev Container, the build process may take a long time
+(as much as 20-40 minutes) as the container collects all the packages of the
```
Note: This should not be the case when this is merged since we will have pre-built images.

max-sixty replied:
Agree, can we remove?

eitsupi replied:
Yes, done.

eitsupi committed:
docs: some updates about VS Code Dev Containers

eitsupi committed:
docs: update title and add note

eitsupi committed:
docs: add a link to containers.dev

eitsupi committed:
docs: formatting lists

pre-commit-ci[bot] committed:
[pre-commit.ci] auto fixes from pre-commit.com hooks

for more information, see https://pre-commit.ci

eitsupi committed:
build: use only Dockerfile for base pre-built image

eitsupi committed:
ci: update workflow to use docker/build-push-action

eitsupi committed:
chore: autoformatting

eitsupi committed:
build: fix workflow trigger

eitsupi committed:
chore: fix path

eitsupi committed:
fix: fix GHA syntax

eitsupi committed:
fix: fix typo

eitsupi committed:
ci: use docker/metadata-action to prepare tags

eitsupi committed:
build: use pre-built image as devcontainer base image

max-sixty commented on book/src/contributing/using-dev-container.md:
```diff
@@ -0,0 +1,55 @@
+1. Clone the git repo onto your hard drive as usual. Then you start the Dev
+   Container (say, using VS Code) that bundles all the developer tools.
+2. Edit files locally (say, using VS Code) to work on PRQL. The tools in the Dev
+   Container watch for changed files and rebuild the project as needed.
```
Is there actually watching going on? I think the volumes are mapped in, but there's no "watch" task, as it would be commonly known as.

(btw this is an example where just linking to docs is easier — can we actually explain this better than VSCode itself can?)

eitsupi replied:
I believe that when the Rust source code changes, rust-analizer automatically builds it.

It is not devcontainer specific, so I agree that it just doesn't need to be explained here.

max-sixty commented on book/src/contributing/using-dev-container.md:
```diff
@@ -0,0 +1,55 @@
+
+## Starting and stopping the Dev Container
+
+_Instructions needed - both from VS Code and maybe from CLI?._
```
I would vote to just link to the standard docs

max-sixty commented on .devcontainer/devcontainer.json:
```diff
@@ -0,0 +1,26 @@
+        "budparr.language-hugo-vscode",
+        "mitsuhiko.insta",
+        "prql-lang.prql-vscode",
+        "esbenp.prettier-vscode"
```
If this supports comments, let's link to https://github.com/PRQL/prql/blob/e132c3de9b1099f0fa0d7929c921a301eaccccc4/.vscode/extensions.json (and link back from there to here) so we keep them all in sync. We should add `rust-analyzer` here too

eitsupi replied:
I think I commented on this before, but did not write it here because the `rust-analizer` is already included in the base image metadata.

However, adding it here is not a problem, so let's add it.

max-sixty commented on .devcontainer/base-image/Dockerfile:
```diff
@@ -0,0 +1,15 @@
+    && rm -rf /var/lib/apt/lists/* \
+    && sh -c "$(curl --location https://taskfile.dev/install.sh)" -- -d -b /usr/local/bin
+
+# ========= Install cargo-tools for non-root user (vscode) =========
```
I'm guessing this is here for good reason — I added a link to the docs, though I couldn't get to a conclusive answer. 

If it still works without switching users, I would vote to not switch users — unless we have a good reason?

```suggestion
# ========= Install cargo-tools for non-root user (vscode) =========
# Based on https://code.visualstudio.com/docs/devcontainers/create-dev-container#_install-additional-software
```

eitsupi replied:
My understanding is that `cargo install` and `task` are usually assumed to be run with user privileges.
So while it may not be necessary to switch from root user here (I may have tried, but I forget), I think it would be more appropriate to do so.

max-sixty replied:
You're correct — they use the `vscode` user upstream... https://github.com/microsoft/vscode-dev-containers/blob/5a084a93b0736ea86395ac99019a5b72a00b6341/containers/rust/.devcontainer/base.Dockerfile

eitsupi replied:
FYI, that is the old definition.
Current definition is this. (The name `vscode` is retained for compatibility)
https://github.com/devcontainers/images/blob/c087e361571e2200c3b40ed86d14a8ced84c388a/src/rust/.devcontainer/devcontainer.json#L9

max-sixty commented on .github/workflows/devcontainer.yaml:
```diff
@@ -0,0 +1,63 @@
+
+on:
+  push:
+    # ToDo: Add partial Taskfile (only include rust-tools)
```
We should indeed add `Taskfile.yml`. 

It's fine to run workflows too often, and we're never going to get it precise, so we should lean on over-building.

max-sixty commented on .github/workflows/devcontainer.yaml:
```diff
@@ -0,0 +1,63 @@
+name: devcontainer
+
+on:
+  push:
```
I think the simplest approach would be to copy the existing workflows like `test-java.yaml` — run on `pull_request`, and then add a call from `test-all.yaml` so it runs on every merge. That removes the need for a `push` action.

Then if something changes that we don't track, we still find out about it. And assuming this is cached, then the build is very fast, so it's fine to run on every merge.

eitsupi replied:
I was simply concerned about the frequent releases of containers.

eitsupi replied:
I recall that frequent builds can be a problem.
That is, even if all the layers that make up this image are the same, the digest of the image itself will be new, making any downstream image cache unavailable.
So we should not build and push the image for every push.

max-sixty replied:
Sorry, I missed this.

If we're concerned about that, we could only push a new image in `release.yaml`.
Or even in `publish-web.yaml`, which would give us the ability to change it faster (we had planned to call that branch `stable`, which would be a more suitable name now...)

But if the files are the same the hash will be the same. But you're correct that the hash can change for a very small reason.

I'll add a TODO to the code — no rush to do this, but if folks reference the workflows it's more cohesive if they follow a similar pattern or have a note where they don't.

eitsupi replied:
That makes sense.
It seems reasonable for me to trigger it when releasing, so I will make that change.

max-sixty commented on Taskfile.yml:
```diff
@@ -52,6 +52,17 @@ tasks:
           🟢 Setup complete! ✅🚀"
         silent: true
 
+  setup-devcontainer:
```
👍

max-sixty commented on book/src/contributing/using-dev-container.md:
````diff
@@ -0,0 +1,55 @@
+
+```admonish note
+This document is a heavy work in progress and may not contain sufficient information.
+```
````
I applaud the approach of disclaiming when something isn't complete, and merging quickly rather than waiting until everything is perfect.

But is there a specific thing these docs lack? To me this looks fairly complete! 

Disclaimers like this should either a) be specific, so we know the next thing to do, or b) have someone's name attached, so they can own getting it to the next stage. Otherwise it's very likely to be there for a long time...

max-sixty commented on .github/workflows/devcontainer.yaml:
```diff
@@ -0,0 +1,63 @@
+
+      - id: set-platforms
+        run: |
+          if [ "${{ github.ref }}" = "refs/heads/main" ]; then
```
Can we try running this on the PR once, so we can test it locally? When I try testing locally, it can't find the image, I guess because it hasn't been pushed yet.

eitsupi replied:
As already commented https://github.com/PRQL/prql/pull/1893#issuecomment-1455409575, I will separate this part into a separate PR and merge it first so that we can try it out.

max-sixty commented on book/src/contributing/using-dev-container.md:
```diff
@@ -0,0 +1,55 @@
+Container command line. Enter one of the following commands to work on a
+component of PRQL.
+
+VS Code offers a link to "Open in Browser" to the proper port to see the results
```
FYI I wasn't familiar with this, and couldn't test it yet — as long as someone is confident then great

eitsupi replied:
I believe VS Code detects when some program uses the port in the container and issues a notification and forwards it to the host machine port. (I don't understand the details of how it works!)

In any case, this will be obvious if anyone use it, so there is little need to specify it here.

@richb-hanover Any thoughts?

richb-hanover replied:
I say, leave out that "open in browser" comment, in the spirit of retaining only what's always going to be true, and not better explained elsewhere.

max-sixty commented on book/src/contributing/using-dev-container.md:
```diff
@@ -0,0 +1,55 @@
+- `task run-playground` starts a Node server to build the Playground. As you
+  edit the files (in the `playground` directory), the server rebuilds those
+  pages. (Port 3000)
+- `task WHAT ELSE?` _Provide explanation of other useful commands._
```
This is actually quite a good list — maybe move to the `developing.md`? Unless there's a specific reason to have this here?

Or put them in `Taskfile.yaml`'s tasks' summary so it's next to the code?

This section should cover just what's _different_ about a dev container.

eitsupi replied:
If the explanation about Taskfile is not elsewhere, it may indeed be better to move it elsewhere.

It should be noted, however, that not all tasks can be executed in all environments.
I was imagining something like a list of tasks available in the container in this page.

max-sixty commented on book/src/contributing/using-dev-container.md:
```diff
@@ -0,0 +1,55 @@
+# Development in Dev Containers
```
FWIW I would match the title of the file with the name of the file

We also need to add this file to the book index

max-sixty reviewed (approved):
Great, code looks v good. 

Some comments about firming up / slimming down the docs — they are well written but it's also important we set them up to improve through time rather than decay — which is more likely if they're concise & specific.

Definitely good to merge and iterate!

eitsupi committed:
Merge branch 'main' into devcontainer

eitsupi committed:
build: remove task for devcontainer setup for now

eitsupi committed:
chore: sync vscode extensions in devcontainer.json

eitsupi committed:
docs: some document update and install zsh completion

pre-commit-ci[bot] committed:
[pre-commit.ci] auto fixes from pre-commit.com hooks

for more information, see https://pre-commit.ci

eitsupi committed:
docs: add the page about devcontainers to the book

eitsupi committed:
build: set postCreateCommand

eitsupi committed:
chore: fix version of go-task Dev Container Feature

max-sixty commented on book/src/contributing/dev-container.md:
````diff
@@ -0,0 +1,46 @@
+# Develop in Dev Containers
+
+```admonish note
+Currently the Dev Container included in this repository only supports the `amd64` platform.
````
Perfect :)

max-sixty commented on book/src/contributing/dev-container.md:
```diff
@@ -0,0 +1,46 @@
+
+[Task](https://taskfile.dev/) is installed in the container for quick access to
+tasks defined on the `Taskfiles.yml`. Autocompression works when using the `zsh`
+shell installed in the container.
```
What is autocompression?

eitsupi replied:
Sorry, it's a typo...... "autocompletion"

The available task suggestions are displayed when the Tab key is pressed.
https://taskfile.dev/installation/#setup-completions

max-sixty replied:
Oh OK, yes that's great. But I wouldn't include that in the PRQL docs...

max-sixty commented on book/src/contributing/dev-container.md:
```diff
@@ -0,0 +1,46 @@
+- `task run-playground` starts a Node server to build the Playground. As you
+  edit the files (in the `playground` directory), the server rebuilds those
+  pages. (Port 3000)
+- `task WHAT ELSE?` _Provide explanation of other useful commands._
```
I would move somewhere else, maybe in a separate PR if @richb-hanover wants to do that. These would be excellent summaries in the Taskfile, and we could just point folks to `task --list` & `task --summary`. 

I was going to suggest https://prql-lang.org/book/contributing/development.html#components-of-prql — but this is out of date now. (which further increases my resolve to ensure we only take docs that we can keep up to date. I'll open an issue for this.)

eitsupi commented on book/src/contributing/dev-container.md:
```diff
@@ -0,0 +1,46 @@
+
+[Task](https://taskfile.dev/) is installed in the container for quick access to
+tasks defined on the `Taskfiles.yml`. Autocompression works when using the `zsh`
+shell installed in the container.
```
```suggestion
tasks defined on the `Taskfiles.yml`.
```

eitsupi committed:
docs: remove superfluous comment

eitsupi committed:
Merge branch 'main' into devcontainer
